### PR TITLE
#488: Add support for Matrix builds

### LINKED ISSUE #488: Support projects with matrix builds
Typical use case: at least 2 version of Exasol database

The following repos currently are using a matrix build

* [spark-connector](https://github.com/exasol/spark-connector/blob/main/.github/workflows/ci-build.yml)
* [cloud-storage-extension](https://github.com/exasol/cloud-storage-extension/blob/main/.github/workflows/ci-build.yml)
* [exasol-testcontainers](https://github.com/exasol/exasol-testcontainers/blob/main/.github/workflows/ci-build.yml)
* [terraform-aws-exasol](https://github.com/exasol/terraform-aws-exasol/blob/main/.github/workflows/ci-build.yml)
* [bucketfs-java](https://github.com/exasol/bucketfs-java/blob/main/.github/workflows/ci-build.yml)
* [row-level-security](https://github.com/exasol/row-level-security/blob/main/.github/workflows/ci-build.yml)
* [remotelog-lua](https://github.com/exasol/remotelog-lua/blob/main/.github/workflows/ci-build.yml)
* [virtual-schema-common-lua](https://github.com/exasol/virtual-schema-common-lua/blob/main/.github/workflows/ci-build.yml)
* [bucketfs-client](https://github.com/exasol/bucketfs-client/blob/main/.github/workflows/ci-build.yml)
* [parquet-edml-generator](https://github.com/exasol/parquet-edml-generator/blob/main/.github/workflows/ci-build.yml)
* [exasol-virtual-schema](https://github.com/exasol/exasol-virtual-schema/blob/main/.github/workflows/ci-build.yml)
* [row-level-security-lua](https://github.com/exasol/row-level-security-lua/blob/main/.github/workflows/ci-build.yml)
* [elasticsearch-virtual-schema](https://github.com/exasol/elasticsearch-virtual-schema/blob/main/.github/workflows/ci-build.yml)
* [exasol-virtual-schema-lua](https://github.com/exasol/exasol-virtual-schema-lua/blob/main/.github/workflows/ci-build.yml)
* [error-reporting-go](https://github.com/exasol/error-reporting-go/blob/main/.github/workflows/ci-build.yml)
* [kinesis-connector-extension](https://github.com/exasol/kinesis-connector-extension/blob/main/.github/workflows/ci-build.yml)
* [exasol-virtual-schema-common-lua](https://github.com/exasol/exasol-virtual-schema-common-lua/blob/main/.github/workflows/ci-build.yml)
* [exasol-driver-ts](https://github.com/exasol/exasol-driver-ts/blob/main/.github/workflows/ci-build.yml)

### PR DESCRIPTION
**Pull request by kaklakariada**
Closes #488

### CODE REVIEW

File: project-keeper/src/main/java/com/exasol/projectkeeper/config/ProjectKeeperRawConfig.java [hunk 1]
```diff
@@ -307,4 +309,53 @@ public void setRelativePath(final String relativePath) {
             this.relativePath = relativePath;
         }
     }
+
+    /**
+     * Intermediate class for reading the CI build configuration sources.
```
Comment (ckunki) [inline project-keeper/src/main/java/com/exasol/projectkeeper/config/ProjectKeeperRawConfig.java:314]:
```suggestion
     * Intermediate class for de-serializing build options from PK's YAML configuration file
     * {@link ProjectKeeperConfigReader.CONFIG_FILE_NAME}.
```

File: project-keeper/src/main/java/com/exasol/projectkeeper/validators/files/CiBuildWorkflowGenerator.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+class CiBuildWorkflowGenerator {
+    private final BuildConfig buildConfig;
+
+    CiBuildWorkflowGenerator(final BuildConfig buildConfig) {
+        this.buildConfig = buildConfig;
+    }
+
+    FileTemplate createReleaseDroidPrepareOriginalChecksumWorkflow() {
+        return new FileTemplateFromResource(".github/workflows/release_droid_prepare_original_checksum.yml",
```
Comment (ckunki) [inline project-keeper/src/main/java/com/exasol/projectkeeper/validators/files/CiBuildWorkflowGenerator.java:16]: We could consider if we want to introduce a builder here as currently the `REQUIRE_EXACT` is followed by some replace operations which reads a little bit strangely in my eyes:

```java 
FileTemplate.builder()
    .fromResource("path/to.yml") // 
    .replacing("a", "b") // 
    .validation(REQUIRE_EXACT) //
    .build();
```

File: project-keeper/src/main/java/com/exasol/projectkeeper/validators/files/FileTemplateFromResource.java [hunk 3]
```diff
@@ -41,14 +41,20 @@ public FileTemplateFromResource(final String templateResource, final String path
     }
 
     /**
-     * Add a replacement definition. {@link FileTemplateFromResource} identifies variables in the template by prefix "$"
-     * and replaces each variable with the value provided beforehand.
+     * Add a replacement definition. {@link FileTemplateFromResource} identifies variables in the template by prefix
+     * {@code $} and replaces each variable with the value provided beforehand.
```
Comment (ckunki) [inline project-keeper/src/main/java/com/exasol/projectkeeper/validators/files/FileTemplateFromResource.java:45]: If I understand correctly, then the curly braces in `if: ${{ $freeDiskSpace }}` will remain unreplaced.
Is this OK (for GH workflows) ?

File: project-keeper/src/main/resources/templates/.github/workflows/ci-build-db-version-matrix.yml [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+        uses: actions/cache@v3
+        with:
+          path: ~/.sonar/cache
+          key: ${{ runner.os }}-sonar
+          restore-keys: ${{ runner.os }}-sonar
+      - name: Enable testcontainer reuse
+        run: echo 'testcontainers.reuse.enable=true' > "$HOME/.testcontainers.properties"
+      - name: Run tests and build with Maven
+        run: |
+          JAVA_HOME=$JAVA_HOME_11_X64 mvn --batch-mode clean verify \
```
Comment (ckunki) [inline project-keeper/src/main/resources/templates/.github/workflows/ci-build-db-version-matrix.yml:50]: Should we add an additional environment variable for Scala?
See https://github.com/exasol/spark-connector/blob/main/.github/workflows/ci-build.yml#L51

**Review by ckunki — CHANGES_REQUESTED** [review]

File: project-keeper/src/main/java/com/exasol/projectkeeper/validators/files/FileTemplateFromResource.java (hunk 3 above)
Comment (kaklakariada) [reply project-keeper/src/main/java/com/exasol/projectkeeper/validators/files/FileTemplateFromResource.java:45]: Yes, that's OK

File: project-keeper/src/main/resources/templates/.github/workflows/ci-build-db-version-matrix.yml (hunk 4 above)
Comment (kaklakariada) [reply project-keeper/src/main/resources/templates/.github/workflows/ci-build-db-version-matrix.yml:50]: Good idea!

File: project-keeper/src/main/java/com/exasol/projectkeeper/validators/files/CiBuildWorkflowGenerator.java (hunk 2 above)
Comment (kaklakariada) [reply project-keeper/src/main/java/com/exasol/projectkeeper/validators/files/CiBuildWorkflowGenerator.java:16]: Good idea! I propose to implement this in a separate PR.

File: project-keeper/src/main/resources/templates/.github/workflows/ci-build-db-version-matrix.yml [hunk 5]
```diff
@@ -51,6 +51,8 @@ jobs:
               -Dorg.slf4j.simpleLogger.log.org.apache.maven.cli.transfer.Slf4jMavenTransferListener=warn \
               -DtrimStackTrace=false \
               -Dcom.exasol.dockerdb.image=${{ matrix.exasol_db_version }}
+        env:
+          EXASOL_DB_VERSION: ${{ matrix.exasol_db_version }}
```
Comment (ckunki) [inline project-keeper/src/main/resources/templates/.github/workflows/ci-build-db-version-matrix.yml:55]:
```suggestion
          # Set additional environment variable as in scala projects the scalatest plugin does not forward
          # the system property -Dcom.exasol.dockerdb.image to the test's implementation.
          EXASOL_DB_VERSION: ${{ matrix.exasol_db_version }}
```

File: shared-model-classes/src/main/java/com/exasol/projectkeeper/shared/config/ProjectKeeperConfig.java [hunk 6]
```diff
@@ -8,46 +8,45 @@
...
     /** @return project version configuration */
     public VersionConfig getVersionConfig() {
-        return versionConfig;
+        return this.versionConfig;
     }
 
-    /** @return CI build runner operating system */
-    public String getCiBuildRunnerOS() {
-        return ciBuildRunnerOS;
+    /** @return CI build configuration */
```
Comment (ckunki) [inline shared-model-classes/src/main/java/com/exasol/projectkeeper/shared/config/ProjectKeeperConfig.java:47]:
```suggestion
    /** @return CI build options */
```

File: shared-model-classes/src/main/java/com/exasol/projectkeeper/shared/config/ProjectKeeperConfig.java [hunk 7]
```diff
@@ -106,12 +105,12 @@ public ProjectKeeperConfig.Builder versionConfig(final VersionConfig versionConf
         }
 
         /**
-         * @param ciBuildRunnerOS operating system for CI builds (default: {@code ubuntu-latest})
+         * @param buildOptions project build configuration
```
Comment (ckunki) [inline shared-model-classes/src/main/java/com/exasol/projectkeeper/shared/config/ProjectKeeperConfig.java:108]:
```suggestion
         * @param buildOptions options for project build workflows
```

File: project-keeper/src/main/java/com/exasol/projectkeeper/config/ProjectKeeperRawConfig.java [hunk 8]
```diff
@@ -103,21 +105,21 @@ public void setVersion(final Object version) {
     }
 
     /**
-     * Get CI build runner operating system, e.g. {@code ubuntu-20.04}.
+     * Get the CI build configuration.
      * 
-     * @return CI build runner operating system
+     * @return build configuration
```
Comment (ckunki) [inline project-keeper/src/main/java/com/exasol/projectkeeper/config/ProjectKeeperRawConfig.java:110]:
```suggestion
     * @return build options
```

File: project-keeper/src/main/java/com/exasol/projectkeeper/config/ProjectKeeperRawConfig.java [hunk 9]
```diff
@@ -103,21 +105,21 @@ public void setVersion(final Object version) {
...
      */
-    public String getCiBuildRunnerOS() {
-        return ciBuildRunnerOS;
+    public Build getBuild() {
+        return build;
     }
 
     /**
-     * Set CI build runner operating system, e.g. {@code ubuntu-20.04}. Default: {@code ubuntu-latest}.
+     * Set the CI build configuration.
```
Comment (ckunki) [inline project-keeper/src/main/java/com/exasol/projectkeeper/config/ProjectKeeperRawConfig.java:117]:
```suggestion
     * Set the CI build options.
```

**Review by ckunki — CHANGES_REQUESTED** [review]

**Review by ckunki — APPROVED** [review]

**Pull request merged**